orb-ui/cone/src/led.rs

```diff
@@ -0,0 +1,140 @@
+use color_eyre::eyre;
+use color_eyre::eyre::Context;
+use ftdi_embedded_hal::eh1::spi::SpiBus;
+use ftdi_embedded_hal::libftd2xx::{Ft4232h, Ftdi, FtdiCommon};
+use orb_rgb::Argb;
+use std::sync::{Arc, Mutex};
+use tokio::sync::{mpsc, watch};
+use tokio::task;
+
+pub struct Led {
+    spi: ftdi_embedded_hal::Spi<Ft4232h>,
+    led_strip_tx: mpsc::UnboundedSender<[Argb; CONE_LED_COUNT]>,
+    task_handle: Option<task::JoinHandle<eyre::Result<()>>>,
+    shutdown_signal: watch::Sender<()>,
+}
+
+pub const CONE_LED_COUNT: usize = 64;
+
+impl Led {
+    pub(crate) fn spawn() -> eyre::Result<Arc<Mutex<Led>>> {
+        let (tx, mut rx) = mpsc::unbounded_channel();
+        let (shutdown_signal, mut shutdown_receiver) = watch::channel(());
+
+        let spi = {
+            let mut device: Ft4232h = Ftdi::with_index(5)?.try_into()?;
+            device.reset().wrap_err("Failed to reset")?;
+            let hal = ftdi_embedded_hal::FtHal::init_freq(device, 3_000_000)?;
+            hal.spi()?
+        };
+
+        let led = Arc::new(Mutex::new(Led {
+            spi,
+            led_strip_tx: tx.clone(),
+            task_handle: None,
+            shutdown_signal,
+        }));
+
+        // spawn receiver thread
+        // where SPI communication happens
+        let led_clone = Arc::clone(&led);
+        let task = task::spawn(async move {
+            loop {
+                // todo do we want to update the LED strip at a fixed rate?
+                // todo do we want to only take the last message and ignore previous ones
+                tokio::select! {
+                    _ = shutdown_receiver.changed() => {
+                        return Ok(());
+                    }
+                    msg = rx.recv() => {
+                        if let Some(msg) = msg {
+                            if let Ok(mut led) = led_clone.lock() {
+                                if let Err(e) = led.spi_rgb_led_update_rgb(&msg) {
+                                    tracing::debug!("Failed to update LED strip: {e}");
+                                }
+                            }
+                        } else {
+                            // none: channel closed
+                            return Err(eyre::eyre!("LED strip receiver channel closed"));
+                        }
+                    }
+                }
+            }
+        });
+
+        if let Ok(led) = &mut led.lock() {
+            led.task_handle = Some(task);
+        }
+
+        tracing::debug!("LED strip initialized");
+
+        Ok(led)
+    }
+
+    pub(crate) fn clone_tx(&self) -> mpsc::UnboundedSender<[Argb; CONE_LED_COUNT]> {
+        self.led_strip_tx.clone()
+    }
+
+    fn spi_rgb_led_update(&mut self, buffer: &[u8]) -> eyre::Result<()> {
```

Comment:
This should take an `impl SpiBus` as an argument. The reason you would use the impl is so that we can provide a mock io Spi device and test this code more easily.

additionally, please provide one unit test by mocking an SpiBus using a vec, and checking that the data written to the vec is as expected.